ui/src/SelectionLayer.tsx

```diff
@@ -0,0 +1,72 @@
+import React from "react";
+import Selection, { Point } from "./Selection";
+import * as uiUtils from "./ui-utils";
+
+interface SelectionCanvasProps {
+  onSelection: (anchor: Point, active: Point) => void;
+}
+
+interface SelectionCanvasState {
+  anchor: Point | null;
+  active: Point | null;
+}
+
+export class SelectionCanvas extends React.Component<
+  SelectionCanvasProps,
+  SelectionCanvasState
+> {
+  constructor(props: SelectionCanvasProps) {
+    super(props);
+    this.state = {
+      anchor: null,
+      active: null
+    };
+  }
+
+  render() {
+    return (
+      <div
+        className="selection-canvas"
+        onClick={e => {
+          if (this.state.anchor === null) {
+            /*
+             * Start selection.
+             */
+            this.setState({ anchor: uiUtils.getMouseXY(e) });
+          } else if (this.state.anchor !== null) {
+            /*
+             * Finalize selection.
+             */
+            if (this.state.active !== null) {
+              this.props.onSelection(this.state.anchor, this.state.active);
+            }
+            this.setState({ anchor: null, active: null });
+          }
+        }}
+        onMouseMove={e => {
+          if (this.state.anchor !== null) {
+            /*
+             * Update selection.
+             */
+            this.setState({ active: uiUtils.getMouseXY(e) });
+          }
+        }}
+        /*
+         * To capture keydown events, tabIndex above must be set to 0.
+         */
+        tabIndex={0}
+        onKeyDown={e => {
+          if (uiUtils.isKeypressEscape(e)) {
+            this.setState({ anchor: null, active: null });
+          }
+        }}
+      >
+        {this.state.anchor !== null && this.state.active !== null && (
+          <Selection anchor={this.state.anchor} active={this.state.active} />
+        )}
+      </div>
+    );
+  }
+}
```

Comment:
Have you tested what happens if the user blurs / defocuses the window?  Things might work as-is, or your mind need to bind to the associated window events to handle them as desired.

Reply:
For the time being, nothing too annoying ^_^

I think the desired behavior is to cancel any selection in-progress when the window is blurred or defocused, and I'll add that in.

A trickier case of undesired behavior is when someone clicks out of a tooltip and it immediately starts a new selection. This feels a bit annoying and I'm trying to think of how to fix it.  One option might be to add a condition to the selection layer that a selection can only start if the selection layer has focus.

In order for that to feel smooth, I would also want to add some indicator that the selection layer has focus, so that on their first click an annotator can see that something has changed, even if their selection doesn't start until they click a second time.